docs/vm-metrics.md

```diff
@@ -0,0 +1,254 @@
+illustrates how the KubeVirt monitoring pipeline will look like
+
+![VM metrics monitoring pipeline diagram](vm-metrics.png "VM Metrics monitoring pipeline")
+
+To have the full set the VM metrics, a guest agent like the QEMU Guest Agent should be run
+in the virtual machines. This step, albeit recommended, is totally left up to the administrator
+of the cluster.
+
+The key components of the monitoring pipeline are:
+
+- The KubeVirt clustr agent: we can use a standard Prometheus server instance
+- The KubeVirt monitoring node agent, see below for the details
+- The KubeVirt HPA API adapters (HPA and historical storage), see below for details
+
+## KubeVirt node agent: collection of VM metrics
+
+The KubeVirt node agent is in charge to connect to the libvirt pod, fetch the VM metrics
+and expose a Prometheus endpoint, to let the KubeVirt cluster agent - a prometheus instance -
+collect and aggregate the cluster metrics.
+
+Optionally, the KubeVirt node agent could do some simple data normalizations as follows:
+
+- The KubeVirt node agent will have one option to always export all the metrics it knows about,
+  even if the real data is not yet available.
+  Such metrics report the special value [TODO: which one? does this
+  concept exist in Kubernetes/Prometheus?] until the real data is available.
+  For example, the VM metric $VM_migration_data_total will be always
+  reported even before the first migration was attempted, if ever.
+
+- The node agent will have one option to keep the last known values of ephemeral metrics
+  after libvirt no longer reports them, for example again the $VM_migration_data_total metric
+  will be reported even after the migration is completed
+
+The normalization features will be controlled by node agent options (e.g. command line switches).
+
+### Metric naming rules
+Exposing the VM metrics to a prometheus server is pretty straightforward considering
+the [prometheus naming rules](https://prometheus.io/docs/practices/naming/).
+We will always one label to identify the VM who exposed the metric, for example
+```
+migration_memory_iteration{vm="$UUID"} 0
+vcpu_total_time{vm="$UUID",cpu="0"} 12345
+```
+
+### Implementation
+
+The requisites of the KubeVirt node agent, besides the obvious need to connect to libvirt and
+to expose a prometheus endpoint, are performance and reliability. The collection of metrics
+should require the minimal amount of resources, both as runtime and as container image.
+We know from the experience that shared storage issues may cause VMs to hang, and libvirt
+may in turn hang while collecting metrics from stuck VMs. We want the Node Agent to cope
+gracefully with this scenario, and keep reporting the metrics of the live VMs, skipping
+the stuck VMs. Reporting of stuck VMs is part of health monitoring, which is out of the
+scope of this document.
+
+The options to implement the KubeVirt node agent are:
+
+#### vAdvisor
```

Comment:
If we really only export prometheus metrics, it might not even be worth to temporarily use vAdvisor. Although I thought it might make sense in the past. Collecting and exposing prometheus metrics is extremely simple in go.